### DISCUSSION TITLE
Parameter learning for BN

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#19] [Q&A]
I think the functions `computeCPTfromDF` found on (https://pyagrum.readthedocs.io/en/1.9.0/notebooks/17-Examples_parametersLearningWithPandas.html)  will throw an error based on the following examples
```
def computeCPTfromDF(bn,df,name):
    """
    Compute the CPT of variable "name" in the BN bn from the database df
    """
    id=bn.idFromName(name)
    parents=list(reversed(bn.cpt(id).names))
    domains=[bn.variableFromName(name).domainSize()
             for name in parents]
    parents.pop()

    if (len(parents)>0):
        c=pd.crosstab(df[name],[df[parent] for parent in parents])
        s=c/c.sum().apply(np.float32)
    else:
        s=df[name].value_counts(normalize=True)

    bn.cpt(id)[:]=np.array((s).transpose()).reshape(*domains)

def ParametersLearning(bn,df):
    """
    Compute the CPTs of every varaible in the BN bn from the database df
    """
    for name in bn.names():
        computeCPTfromDF(bn,df,name)
```

**Examples**

```

bn=gum.fastBN("A[10]->B[10]<-C[10]")
print(bn.cpt("A")* bn.cpt("B") * bn.cpt("C"))
gum.generateSample(bn, 100, name_out ="random.csv")
df = pd.read_csv("random.csv")
ParametersLearning(bn,df)
print(bn.cpt("A")* bn.cpt("B") * bn.cpt("C"))
```

**Comment by phwuil** [comment]
You are right but I am not familiar enough with pandas to correct it :-)

The problem is that you generate only *100* samples for a joint of size  `10x10x10` => obvisouly, you do not have all the cases for the joint.  Then the crosstab does not contain all the cases and has a size < 1000. Then you cannot reshape it.

For instance, with only 10 generated samples, here is the crosstab :
![image](https://github.com/agrumery/aGrUM/assets/5980636/074676db-08f5-447d-8fe5-5b8e2fb29cb1)

As you can see, `B=1` is never seen, `(A,C)` in {`(0,0)`,`(0,1)`,`(1,0)`,...} neither ...  So the crosstab have a size 70 that we want to reshape in `10x10x10`. Leading to the error :

![image](https://github.com/agrumery/aGrUM/assets/5980636/6357c325-1f1b-4438-b21b-19e2087690ae)


If you generate enough sample (say 10 000), the generation explore all the possible events and then the crosstab contains 10x10x10 elements and every thing is OK :-)

The solution should be to add columns and rows of 0.0 for all the missing event... But I do not know how to do that in pandas

**Reply by phwuil** [reply]
Note that it is much more convenient to use `pyAgrum.BNLearner.fitParameters` :-)

![image](https://github.com/agrumery/aGrUM/assets/5980636/223c3c88-dce2-4cb9-9c8a-f458e126eb18)

But you stil have the problem with a too small sample size

![image](https://github.com/agrumery/aGrUM/assets/5980636/3cd28d2a-d196-43f9-ab21-5549eba3f581)

Except if you use use a prior or a smoothing as proposed in the error above :

![image](https://github.com/agrumery/aGrUM/assets/5980636/36976856-706c-4d54-b50d-ff7f03a20433)

**Comment by kenneth-lee-ch** [comment]
What is the difference between the functions I posted and the
pyAgrum.BNLearner.fitParameters?

**Reply by phwuil** [reply]
`fitParameters` does not use panda. This notebook is just an example to understand how parameter learning works but it is not a correct way to do it. 

`fitParameters` (or the less convenient `learnParameters`) can include prior or smoothing, is multithreaded and written in C++ to be as fast as possible.

**Comment by kenneth-lee-ch** [comment]
So, if I don’t use prior smoothing, the resulting potential will be the same?

**Reply by phwuil** [reply]
It should :-)

But, once again , for your example of 100 samples, there is no way to fit the parameters correctly without priors or smoothing

**Comment by kenneth-lee-ch** [comment]
But I guess using fit.parameter will automatically fill in 0 for the missing events?

**Reply by phwuil** [reply]
The problem is when a conditionning set is not observed.
Let say that `A[3]` is never observed with the value 1.

` P(A=1)=0`  can be correctly computed from the samples. (but is this correct ? proba=0 means **never** )

but if `A->B` in the BN, how can we computed `P(B|A=1)` since we never observed values of B when A is 1 ... 
`P(B|A=1)=[0,0]` ...

- With smoothing, `P(B|A=1)=[0.5 , 0.5 ]`
- With prior (Dirichlet for instance) you can specify more precisely this distribution
